# Importing libraries.

In [1]:
import os
import pandas as pd

from sklearn.ensemble import RandomForestClassifier

# Acquiring the data.

In [2]:
df_train = pd.read_csv('/kaggle/input/titanic/train.csv')
df_test = pd.read_csv('/kaggle/input/titanic/test.csv')

#Performing concatenation on the dataframes to ease further manipulations.
def concatenation(df_train, df_test):
    return(pd.concat([df_train, df_test]).reset_index(drop = True))

#Performing division on the concatenated dataframes.
def division(df_all):
    return(df_all.iloc[:890], df_all.iloc[891:].drop(['Survived'], axis = 1))

df_all = concatenation(df_train, df_test)

df_all.sample(5)

PassengerId  Survived  Pclass  \
911          912       NaN       1   
260          261       0.0       3   
192          193       1.0       3   
24            25       0.0       3   
257          258       1.0       1   

                                                Name     Sex   Age  SibSp  \
911                           Rothschild, Mr. Martin    male  55.0      1   
260                                Smith, Mr. Thomas    male   NaN      0   
192  Andersen-Jensen, Miss. Carla Christine Nielsine  female  19.0      1   
24                     Palsson, Miss. Torborg Danira  female   8.0      3   
257                             Cherry, Miss. Gladys  female  30.0      0   

     Parch    Ticket     Fare Cabin Embarked  
911      0  PC 17603  59.4000   NaN        C  
260      0    384461   7.7500   NaN        Q  
192      0    350046   7.8542   NaN        S  
24       1    349909  21.0750   NaN        S  
257      0    110152  86.5000   B77        S

# Analysis of the data.

**Analysis of the dataframes.**

In [3]:
df_train.info()
print('\n')

for column in df_train:
    print(f'{column} contains {df_train[column].isna().sum()} missing values.')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


PassengerId contains 0 missing values.
Survived contains 0 missing values.
Pclass contains 0 missing values.
Name contains 0 missing values.
Sex contains 0 missing values.
Age contains 177 missing values.
SibSp contains 0 missing values.
Parc

Train dataframe contains:
* 891 rows, 12 features.
* 3 columns with missing values, such as *Age* (177), *Cabin* (687) and *Embarked* (2).
* 7 columns of integers or float type.
* 5 columns of object (string) type.

In [4]:
df_test.info()
print('\n')

for column in df_test:
    print(f'{column} contains {df_test[column].isna().sum()} missing values.')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  418 non-null    int64  
 1   Pclass       418 non-null    int64  
 2   Name         418 non-null    object 
 3   Sex          418 non-null    object 
 4   Age          332 non-null    float64
 5   SibSp        418 non-null    int64  
 6   Parch        418 non-null    int64  
 7   Ticket       418 non-null    object 
 8   Fare         417 non-null    float64
 9   Cabin        91 non-null     object 
 10  Embarked     418 non-null    object 
dtypes: float64(2), int64(4), object(5)
memory usage: 36.0+ KB


PassengerId contains 0 missing values.
Pclass contains 0 missing values.
Name contains 0 missing values.
Sex contains 0 missing values.
Age contains 86 missing values.
SibSp contains 0 missing values.
Parch contains 0 missing values.
Ticket contains 0 missing values.
Fare contains 1 

Test dataframe also contains:
* 418 rows, 11 features.
* 3 columns with missing values, such as *Age* (86), *Fare* (1) and *Cabin* (327).
* 6 columns of integers or float type.
* 5 columns of object (string) type.

**Analysis of the train dataframe.**

In [5]:
df_train.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [6]:
df_train.describe(include = ['O'])

Name   Sex  Ticket    Cabin Embarked
count                       891   891     891      204      889
unique                      891     2     681      147        3
top     Braund, Mr. Owen Harris  male  347082  B96 B98        S
freq                          1   577       7        4      644

**Analysis of the test dataframe.**

In [7]:
df_test.describe()

PassengerId      Pclass         Age       SibSp       Parch        Fare
count   418.000000  418.000000  332.000000  418.000000  418.000000  417.000000
mean   1100.500000    2.265550   30.272590    0.447368    0.392344   35.627188
std     120.810458    0.841838   14.181209    0.896760    0.981429   55.907576
min     892.000000    1.000000    0.170000    0.000000    0.000000    0.000000
25%     996.250000    1.000000   21.000000    0.000000    0.000000    7.895800
50%    1100.500000    3.000000   27.000000    0.000000    0.000000   14.454200
75%    1204.750000    3.000000   39.000000    1.000000    0.000000   31.500000
max    1309.000000    3.000000   76.000000    8.000000    9.000000  512.329200

In [8]:
df_test.describe(include = ['O'])

Name   Sex    Ticket            Cabin Embarked
count                418   418       418               91      418
unique               418     2       363               76        3
top     Kelly, Mr. James  male  PC 17608  B57 B59 B63 B66        S
freq                   1   266         5                3      270

# Filling the missing values.

***Age* column.**

Instead of filling the column with mean, median or any other value regarding the whole dataset, I will fill the missing values with median age for passenger class and sex.

In [9]:
print(df_all.groupby(['Sex', 'Pclass'])['Age'].median())

df_all['Age'] = df_all.groupby(['Sex', 'Pclass'])['Age'].apply(lambda x: x.fillna(value = x.median()))

Sex     Pclass
female  1         36.0
        2         28.0
        3         22.0
male    1         42.0
        2         29.5
        3         25.0
Name: Age, dtype: float64


***Embarked* column.**

Since there are only 2 missing values in the *Embarked* column, I will display them and try to look up for something interesting. As shown below, 2 rows with missing values have the same ticket, fare and cabin, I conclude that these 2 passengers was together and embarked from the same place. Let's just fill the missing values with the most frequent *Embarked* value, which is 'S'.

In [10]:
df_all[df_all['Embarked'].isna()]

PassengerId  Survived  Pclass                                       Name  \
61            62       1.0       1                        Icard, Miss. Amelie   
829          830       1.0       1  Stone, Mrs. George Nelson (Martha Evelyn)   

        Sex   Age  SibSp  Parch  Ticket  Fare Cabin Embarked  
61   female  38.0      0      0  113572  80.0   B28      NaN  
829  female  62.0      0      0  113572  80.0   B28      NaN

In [11]:
df_all['Embarked'] = df_all['Embarked'].fillna(value = 'S')

***Fare* column.**

Only 1044 passenger has a missing value in the *Fare* column. I will fill it with a median value for his passenger class.

In [12]:
df_all[df_all['Fare'].isna()]

PassengerId  Survived  Pclass                Name   Sex   Age  SibSp  \
1043         1044       NaN       3  Storey, Mr. Thomas  male  60.5      0   

      Parch Ticket  Fare Cabin Embarked  
1043      0   3701   NaN   NaN        S

In [13]:
print(df_all.groupby(['Pclass'])['Fare'].median())

df_all['Fare'] = df_all.groupby(['Pclass'])['Fare'].apply(lambda x: x.fillna(value = x.median()))

Pclass
1    60.0000
2    15.0458
3     8.0500
Name: Fare, dtype: float64


I will not fill the *Cabin* column, because it contains too many missing values.

# Feature engineering and data wrangling.

First off, let's drop *Ticket* and *Cabin* columns. Then I will convert categorical values (*Sex* and *Embarked* columns) into a numerical values.

In [14]:
df_all = df_all.drop(['Ticket', 'Cabin'], axis = 1)

df_all['Sex'] = df_all['Sex'].map({'male': 1, 'female': 0})
df_all['Embarked'] = df_all['Embarked'].map({'C': 0, "Q": 1, 'S': 2})

df_all.sample(5)

PassengerId  Survived  Pclass  \
325          326       1.0       1   
569          570       1.0       3   
436          437       0.0       3   
617          618       0.0       3   
835          836       1.0       1   

                                                Name  Sex   Age  SibSp  Parch  \
325                         Young, Miss. Marie Grice    0  36.0      0      0   
569                                Jonsson, Mr. Carl    1  32.0      0      0   
436             Ford, Miss. Doolina Margaret "Daisy"    0  21.0      2      2   
617  Lobb, Mrs. William Arthur (Cordelia K Stanlick)    0  26.0      1      0   
835                      Compton, Miss. Sara Rebecca    0  39.0      1      1   

         Fare  Embarked  
325  135.6333         0  
569    7.8542         2  
436   34.3750         2  
617   16.1000         2  
835   83.1583         0

Secondly, let's extract titles from the *Name* column and convert them into numerical values.

In [15]:
#Extracting the titles.
df_all['Title'] = df_all.Name.str.extract(' ([A-Za-z]+)\.', expand = False)

print(df_all.groupby(['Title'])['Survived'].mean().sort_values(ascending = False), '\n')

#Replacing odd titles with more usual titles.
df_all['Title'] = df_all['Title'].replace(['Lady', 'Countess','Capt', 'Col','Don', 'Dr', 'Major', 'Rev', 'Sir', 'Jonkheer', 'Dona'], 'Rare titles')
df_all['Title'] = df_all['Title'].replace('Mlle', 'Miss')
df_all['Title'] = df_all['Title'].replace('Ms', 'Miss')
df_all['Title'] = df_all['Title'].replace('Mme', 'Mrs')

df_all['Title'] = df_all['Title'].map({'Mrs': 0, 'Miss': 1, 'Master': 2, 'Rare titles': 3, 'Mr': 4})

print(df_all.groupby(['Title'])['Survived'].mean().sort_values(ascending = False), '\n')

df_all = df_all.drop(['Name'], axis = 1)

df_all.sample(5)

Title
Sir         1.000000
Countess    1.000000
Ms          1.000000
Mme         1.000000
Lady        1.000000
Mlle        1.000000
Mrs         0.792000
Miss        0.697802
Master      0.575000
Col         0.500000
Major       0.500000
Dr          0.428571
Mr          0.156673
Jonkheer    0.000000
Don         0.000000
Rev         0.000000
Capt        0.000000
Dona             NaN
Name: Survived, dtype: float64 

Title
0    0.793651
1    0.702703
2    0.575000
3    0.347826
4    0.156673
Name: Survived, dtype: float64 



PassengerId  Survived  Pclass  Sex   Age  SibSp  Parch     Fare  \
1152         1153       NaN       3    1  21.0      0      0   7.8542   
372           373       0.0       3    1  19.0      0      0   8.0500   
391           392       1.0       3    1  21.0      0      0   7.7958   
759           760       1.0       1    0  33.0      0      0  86.5000   
766           767       0.0       1    1  42.0      0      0  39.6000   

      Embarked  Title  
1152         2      4  
372          2      4  
391          2      4  
759          2      3  
766          0      3

Thirdly, let's redefine ages as age intervals.

In [16]:
df_all['Age intervals'] = pd.cut(df_all['Age'], 4)

print(df_all.groupby(['Age intervals'])['Survived'].mean().sort_values(ascending = False), '\n')

df_all.loc[df_all['Age'] <= 16, 'Age'] = 0
df_all.loc[(df_all['Age'] > 16) & (df_all['Age'] <= 32), 'Age'] = 1
df_all.loc[(df_all['Age'] > 32) & (df_all['Age'] <= 48), 'Age'] = 2
df_all.loc[(df_all['Age'] > 48) & (df_all['Age'] <= 64), 'Age'] = 3
df_all.loc[df_all['Age'] > 64, 'Age'] = 4

df_all = df_all.drop(['Age intervals'], axis = 1)

df_all.sample(5)

Age intervals
(0.0902, 20.128]    0.458101
(20.128, 40.085]    0.369686
(40.085, 60.043]    0.369128
(60.043, 80.0]      0.227273
Name: Survived, dtype: float64 



PassengerId  Survived  Pclass  Sex  Age  SibSp  Parch     Fare  Embarked  \
741          742       0.0       1    1  2.0      1      0  78.8500         2   
575          576       0.0       3    1  1.0      0      0  14.5000         2   
681          682       1.0       1    1  1.0      0      0  76.7292         0   
405          406       0.0       2    1  2.0      1      0  21.0000         2   
479          480       1.0       3    0  0.0      0      1  12.2875         2   

     Title  
741      4  
575      4  
681      4  
405      4  
479      1

Now, let's create a new column *Family* which contains the sum of *SibSp* and *Parch* columns.

In [17]:
df_all['Family'] = df_all['SibSp'] + df_all['Parch']

df_all = df_all.drop(['SibSp', 'Parch'], axis = 1)

print(df_all.groupby(['Family'])['Survived'].mean().sort_values(ascending = False), '\n')

df_all.sample(5)

Family
3     0.724138
2     0.578431
1     0.552795
6     0.333333
0     0.303538
4     0.200000
5     0.136364
7     0.000000
10    0.000000
Name: Survived, dtype: float64 



PassengerId  Survived  Pclass  Sex  Age     Fare  Embarked  Title  \
1110         1111       NaN       3    1  1.0   8.0500         2      4   
453           454       1.0       1    1  3.0  89.1042         0      4   
878           879       0.0       3    1  1.0   7.8958         2      4   
1019         1020       NaN       2    1  2.0  13.0000         2      4   
60             61       0.0       3    1  1.0   7.2292         0      4   

      Family  
1110       0  
453        1  
878        0  
1019       0  
60         0

Lastly, let's replace the *Fare* column values with categorical values.

In [18]:
df_all['Fare intervals'] = pd.qcut(df_all['Fare'], 5)

print(df_all.groupby(['Fare intervals'])['Survived'].mean().sort_values(ascending = False), '\n')

df_all.loc[df_all['Fare'] <= 7, 'Fare'] = 0
df_all.loc[(df_all['Fare'] > 7) & (df_all['Fare'] <= 14), 'Fare'] = 1
df_all.loc[(df_all['Fare'] > 14) & (df_all['Fare'] <= 31), 'Fare'] = 2
df_all.loc[df_all['Fare'] > 31, 'Fare'] = 3

df_all = df_all.drop(['Fare intervals'], axis = 1)

Fare intervals
(41.579, 512.329]    0.645349
(21.558, 41.579]     0.443243
(10.5, 21.558]       0.426901
(-0.001, 7.854]      0.217877
(7.854, 10.5]        0.201087
Name: Survived, dtype: float64 



Let's look at the dataframe we have created.

In [19]:
df_all.sample(10)

PassengerId  Survived  Pclass  Sex  Age  Fare  Embarked  Title  Family
9              10       1.0       2    0  0.0   2.0         0      0       1
1255         1256       NaN       1    0  1.0   3.0         0      0       1
526           527       1.0       2    0  3.0   1.0         2      1       0
725           726       0.0       3    1  1.0   1.0         2      4       0
1049         1050       NaN       1    1  2.0   2.0         2      4       0
116           117       0.0       3    1  4.0   1.0         1      4       0
296           297       0.0       3    1  1.0   1.0         0      4       0
970           971       NaN       3    0  1.0   1.0         1      1       0
1184         1185       NaN       1    1  3.0   3.0         2      3       2
1004         1005       NaN       3    0  1.0   1.0         1      1       0

# Assigning the data to x and y values and performing the RandomForest algorithm.

In [20]:
df_train, df_test = division(df_all)

x_train = df_train.drop(['Survived'], axis = 1)
y_train = df_train['Survived']

x_test = df_test

In [21]:
model = RandomForestClassifier(criterion = 'gini',
                               n_estimators = 2000,
                               max_depth = 5,
                               max_features = 'sqrt',
                               random_state = 20)
model.fit(x_train, y_train)
y_pred = model.predict(x_test).astype('int64')

# Making a submission.

In [22]:
submission = pd.DataFrame({'PassengerId': df_test['PassengerId'], 'Survived': y_pred})
submission.to_csv('submission.csv', index = False)